### ISSUE TITLE
Segfault in MainWindow::delete_old_pictures

### ISSUE THREAD

**Issue by 414n** [pagaco-swita/ultimatedailywallpaper#1]
Greetings,

yesterday I started noticing that the program does not work anymore on my system (Slackware64-15.0 Linux with KDE/Plasma 5).
When launching it from a terminal, I noticed a SEGFAULT error that prevented it from working normally.

I recompiled the project with debug flags and I got this backtrace from gdb:

```gdb
Starting program: /usr/bin/UltimateDailyWallpaper 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0x7ffff3062640 (LWP 16717)]
[New Thread 0x7ffff08b8640 (LWP 16718)]
[New Thread 0x7fffeb5ef640 (LWP 16719)]
[New Thread 0x7fffeac4a640 (LWP 16720)]
[New Thread 0x7fffea449640 (LWP 16721)]
[Thread 0x7fffeac4a640 (LWP 16720) exited]
[New Thread 0x7fffeac4a640 (LWP 16722)]
[New Thread 0x7fffd3cfa640 (LWP 16723)]
[New Thread 0x7fffd335c640 (LWP 16724)]
[New Thread 0x7fffd2b5b640 (LWP 16725)]
[New Thread 0x7fffd235a640 (LWP 16726)]
[Detaching after fork from child process 16727]
[Detaching after fork from child process 16730]
[New Thread 0x7fffd1580640 (LWP 16732)]
[Thread 0x7fffd1580640 (LWP 16732) exited]

Thread 1 "UltimateDailyWa" received signal SIGSEGV, Segmentation fault.
0x00007ffff6bf1793 in QString::append(QString const&) () from /usr/lib64/libQt5Core.so.5
(gdb) bt
#0  0x00007ffff6bf1793 in QString::append(QString const&) () from /usr/lib64/libQt5Core.so.5
#1  0x000000000040e1fb in QString::operator+= (this=0x7fffffffd6e8, s=...) at /usr/include/qt5/QtCore/qstring.h:569
#2  0x000000000040e387 in operator+ (s1=..., s2=...) at /usr/include/qt5/QtCore/qstring.h:1531
#3  0x000000000041b8a4 in MainWindow::delete_old_pictures (this=0x7fffffffdd60) at mainwindow.cpp:1159
#4  0x00000000004154bd in MainWindow::check_content (this=0x7fffffffdd60) at mainwindow.cpp:540
#5  0x00000000004142c3 in MainWindow::MainWindow (this=0x7fffffffdd60, parent=0x0) at mainwindow.cpp:471
#6  0x0000000000411335 in main (argc=1, argv=0x7fffffffe0a8) at main.cpp:26
(gdb) bt full
#0  0x00007ffff6bf1793 in QString::append(QString const&) () from /usr/lib64/libQt5Core.so.5
No symbol table info available.
#1  0x000000000040e1fb in QString::operator+= (this=0x7fffffffd6e8, s=...) at /usr/include/qt5/QtCore/qstring.h:569
No locals.
#2  0x000000000040e387 in operator+ (s1=..., s2=...) at /usr/include/qt5/QtCore/qstring.h:1531
        t = {d = 0x5ed100}
#3  0x000000000041b8a4 in MainWindow::delete_old_pictures (this=0x7fffffffdd60) at mainwindow.cpp:1159
        oldwallfile = {d = 0x7ffff6e21a80 <QArrayData::shared_null>}
        old_wallp = <incomplete type>
        oldthumbfile = {d = 0x5ed100}
        old_thumbf = <incomplete type>
        j = 4
#4  0x00000000004154bd in MainWindow::check_content (this=0x7fffffffdd60) at mainwindow.cpp:540
No locals.
#5  0x00000000004142c3 in MainWindow::MainWindow (this=0x7fffffffdd60, parent=0x0) at mainwindow.cpp:471
        _settDir = {d_ptr = {d = 0x61d540}}
        _locale = {d = 0x5f5690}
        _loadpath = {d = 0x608b70}
        _confDirVariable = {d = 0x612d90}
        langfilename = {d = 0x5ee840}
        _autorun_location = {d = 0x6213c0}
        _detected_autorun_dirs = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {
                static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, 
                          _M_i = -1}, static is_always_lock_free = true}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, 
                d = 0x7ffff6e22590 <QListData::shared_null>}, d = 0x7ffff6e22590 <QListData::shared_null>}}, <No data fields>}
        confFile = <incomplete type>
        _current_desktop = {d = 0x60e860}
        autorun_file = <incomplete type>
        _autorundir = {d_ptr = {d = 0x602570}}
        _settDir = <optimized out>
        confFile = <optimized out>
        _current_desktop = <optimized out>
        _locale = <optimized out>
        _loadpath = <optimized out>
        langfilename = <optimized out>
        _confDirVariable = <optimized out>
        _autorun_location = <optimized out>
        _detected_autorun_dirs = <optimized out>
        autorun_file = <optimized out>
        _autorundir = <optimized out>
        MsgError = <optimized out>
        langfilesdir = <optimized out>
        _defaultcontent = <optimized out>
        reply = <optimized out>
        stream = <optimized out>
        kdefullsession = <optimized out>
        kdeversion = <optimized out>
        _autorun_file_content = <optimized out>
        stream = <optimized out>
        MsgError = <optimized out>
#6  0x0000000000411335 in main (argc=1, argv=0x7fffffffe0a8) at main.cpp:26
        a = <incomplete type>
        w = {<QMainWindow> = {<No data fields>}, static staticMetaObject = {d = {superdata = {
                direct = 0x7ffff7f9fea0 <QMainWindow::staticMetaObject>}, 
              stringdata = 0x4561c0 <qt_meta_stringdata_MainWindow>, data = 0x4566c0 <qt_meta_data_MainWindow>, 
              static_metacall = 0x433646 <MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, 
              relatedMetaObjects = 0x0, extradata = 0x0}}, setwall = {<QObject> = {<No data fields>}, 
            static staticMetaObject = {d = {superdata = {direct = 0x7ffff6fd50e0 <QObject::staticMetaObject>}, 
                stringdata = 0x458540 <qt_meta_stringdata_setWallpaper>, data = 0x458680 <qt_meta_data_setWallpaper>, 
                static_metacall = 0x434626 <setWallpaper::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, 
                relatedMetaObjects = 0x0, extradata = 0x0}}, out = {d = 0x7ffff6e21a80 <QArrayData::shared_null>}, 
            _detected_monitors = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {
                    d = 0x7ffff6e22590 <QListData::shared_null>}, 
                  d = 0x7ffff6e22590 <QListData::shared_null>}}, <No data fields>}}, basicinterface = 0x6216f0, 
          extendedfunctioninterface = 0x0, additionalfunctioninterface = 0x2, menuinterface = 0x0, submenuinterface = 0x0, 
          translator = 0x5ede60, conn = {d_ptr = 0x9558f0}, mSystemTrayIcon = 0x5f12b0, _autoChangeTimer = 0x955860, 
          detected_plugins = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x6485f0}, 
                d = 0x6485f0}}, <No data fields>}, 
          detected_providers = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x5f2ce0}, 
                d = 0x5f2ce0}}, <No data fields>}, 
          datelist = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x9aca30}, 
                d = 0x9aca30}}, <No data fields>}, 
          selected_datelist = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x9a1c10}, 
                d = 0x9a1c10}}, <No data fields>}, 
          filenamelist = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x9a0360}, 
                d = 0x9a0360}}, <No data fields>}, 
          _oldthumbfiles = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x988dc0}, 
                d = 0x988dc0}}, <No data fields>}, 
          _record_out = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {d = 0x621ec0}, 
                d = 0x621ec0}}, <No data fields>}, _AppFileDir = {d = 0x624af0}, _confFile = {d = 0x5d9460}, _picturedir = {
            d = 0x626490}, _lang = {d = 0x99f7e0}, _databaseFilePath = {d = 0x6155e0}, _thumbfiledir = {d = 0x61da70}, 
          _selected_provider = {d = 0x629a80}, _selected_plugin = {d = 0x623b10}, _pluginsdir = {d = 0x6252e0}, 
          _langfilesdir = {d = 0x612530}, _copyright = {d = 0x7ffff6e21a80 <QArrayData::shared_null>}, _applang = {
            d = 0x5f2770}, _last_change = {d = 0x614180}, plugdir = {d = 0x7ffff6e21a80 <QArrayData::shared_null>}, 
          _last_file = {d = 0x622230}, _screenCurrentlyLocked = false, _needsRefreshAfterUnlock = false, 
          regular_request = true, _store_days = 15, _autochange = 2, _used_desktop = 6, _time_hours = 11, _time_minutes = 0, 
          _autorun = 2, _change_lockscreen_wallpaper = 2, _headImage = <incomplete type>, _descImage = <incomplete type>, 
          _descWidget = 0x624270, dL = 0x614d10, _widgetaction = 0x968b90, _imageLabel = 0x6a3980, _labelTitle = 0x6a32e0, 
          _labelDescription = 0x61b850, basemenu = 0x9a00b0, provider = 0x8fa4c0, 
          getnewpicture = 0x4330b1 <__static_initialization_and_destruction_0(int, int)+79>, moreinformation = 0x8c5930, 
          loadexistingpicture = 0x850690, settings = 0x8c5830, aboutapp = 0x8fa3d0, aboutplugin = 0x2, quitapp = 0x8fa3f0, 
          wmc_potd_morepictures = 0x7ffff6617e48 <__exit_funcs_lock>, 
          provider_wikimedia_commons_potd = 0x434920 <__libc_csu_init>}
```

From the looks of it, there seems to be a problem in `MainWindow::delete_old_pictures`, when declaring `oldthumbfile`

**Comment by pagaco-swita** [comment]
Hello 414n

Thank you for filing this bug for the detailed information!
I have determined the cause of the error.
The lists with the filenames __oldthumbfiles_ and _filenamelist_ were not created correctly in MainWindow::create_filenamelist(). This then led to a crash at that point in _MainWindow::delete_old_pictures_.
I thought I checked this function, but I must have missed it.
I just fixed the bug and released version v3.3.5.

Best regards,
Patrice Coni
